### ISSUE TITLE
Schema.getAdditionalProperties(): return something better as Object?

### ISSUE THREAD

**Issue by jmini** [eclipse/microprofile-open-api#257]
Per spec: `additionalProperties` in a Schema object can be an Object or a Boolean.
See https://github.com/OAI/OpenAPI-Specification/issues/668#issuecomment-281837896

In the code, `Object` is returned:
https://github.com/eclipse/microprofile-open-api/blob/c2a26360d460bfe07b30c83ba4fe6f9cc73543c6/api/src/main/java/org/eclipse/microprofile/openapi/models/media/Schema.java#L641

This is not really user-friendly because the user needs to know that this `Object` can be a `Schema` or a `Boolean` and write code with `instanceOf` to reflect this.

---

Kaizen Model has 2 getters for this:

* `Boolean getAdditionalProperties()`
* `Schema getAdditionalPropertiesSchema()`

I do not like this either, but it is better in a sense that no `instanceOf` is needed.

[kaizen-parser Schema.java](https://github.com/RepreZen/KaiZen-OpenApi-Parser/blob/9295db46b47406aaa7046ceca5ef03d331cd6a51/kaizen-openapi-parser/src/main/java/com/reprezen/kaizen/oasparser/model3/Schema.java#L314-L332)

---

What is the typical java way of telling `Schama` or `Boolean` (a sort of Union Types)?
A new Interface `AdditionalProperties` with 2 getters returning an optional: 

* `Optional<Schema> getAsSchema()` 
* `Optional<Boolean> getAsBoolean()`

**Comment by jmini** [comment]
Thank you for discussing this in the Hangout, and thank you for reporting here the decision.

Current (1.0):
```
Object getAdditionalProperties()
```

---

Is the idea to have:

```
Boolean getAdditionalPropertiesBoolean()
Schema getAdditionalPropertiesSchema()
```

Then I understand that `getAdditionalProperties()` needs to be removed with 2.0 (and can be deprecated with 1.1).

I prefer this solution, because if users are using `Object o = s.getAdditionalProperties()` it is not clear that the method now only returns Boolean.

---

But you have mentioned:

```
Boolean getAdditionalProperties()
Schema getAdditionalPropertiesSchema()
```

Then the only modification is that `getAdditionalProperties()` was returning an `Object` with 1.0 and will return a `Boolean` with 2.0.

**Comment by jmini** [comment]
Additional proposition:

We have also already 2 setters:
https://github.com/eclipse/microprofile-open-api/blob/8e0f2479a00470911d64ae13a59c5cd78f4240c9/api/src/main/java/org/eclipse/microprofile/openapi/models/media/Schema.java#L643-L659

I think they should be renamed to:
```java
Boolean setAdditionalPropertiesBoolean(Boolean)
Schema setAdditionalPropertiesSchema(Schema)
```

In order to be consistent with the getters.

**Comment by arthurdm** [comment]
Good catch @jmini 

We discussed in today's hangout and agreed with the additions of:
```
Boolean getAdditionalPropertiesBoolean()
Schema getAdditionalPropertiesSchema()
Boolean setAdditionalPropertiesBoolean(Boolean)
Schema setAdditionalPropertiesSchema(Schema)
```

And deprecate the existing ones in `1.1`, removing then in `2.0`

**Comment by jmini** [comment]
PR for this: #277. Feedback welcomed.

**Comment by jmini** [comment]
PR for `2.0`: #281

**Comment by jmini** [comment]
Both PRs has been merged (for `1.1` and `2.0`). Nothing more to do in my opinion, I am closing this issue. Thank you a lot.

**Comment by jmini** [comment]
Both PRs has been merged (for `1.1` and `2.0`). Nothing more to do in my opinion, I am closing this issue. Thank you a lot.
